src/tlv.rs

```diff
@@ -0,0 +1,252 @@
+//! BER-TLV writer and parser
+
+use crate::command::{writer::Error as _, DataSource, DataStream, Writer};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+pub struct Tag([u8; 3]);
+
+impl Tag {
+    pub const fn from_u8(value: u8) -> Self {
+        Tag([0, 0, value])
+    }
+
+    pub const fn from_u16(value: u16) -> Self {
+        Tag::from_2(value.to_be_bytes())
+    }
+
+    pub const fn from_2([b1, b2]: [u8; 2]) -> Self {
+        if b1 == 0 {
+            Tag([0, 0, b2])
+        } else {
+            Tag([0, b1, b2])
+        }
+    }
+    pub const fn from_3([b1, b2, b3]: [u8; 3]) -> Self {
+        if b1 == 0 {
+            Self::from_2([b2, b3])
+        } else {
+            Tag([b1, b2, b3])
+        }
+    }
+}
+
+impl From<u8> for Tag {
+    fn from(value: u8) -> Self {
+        Self::from_u8(value)
+    }
+}
+
+impl From<u16> for Tag {
+    fn from(value: u16) -> Self {
+        value.to_be_bytes().into()
+    }
+}
+
+impl From<[u8; 1]> for Tag {
+    fn from([value]: [u8; 1]) -> Self {
+        value.into()
+    }
+}
+
+impl From<[u8; 2]> for Tag {
+    fn from(value: [u8; 2]) -> Self {
+        Self::from_2(value)
+    }
+}
+
+impl From<[u8; 3]> for Tag {
+    fn from(value: [u8; 3]) -> Self {
+        Self::from_3(value)
+    }
+}
+
+impl Tag {
+    pub fn serialize(&self) -> heapless::Vec<u8, 3> {
+        let [b1, b2, b3] = self.0;
+        if b1 == 0 {
+            if b2 == 0 {
+                debug_assert_ne!(b3 & 0b11111, 0b11111, "Invalid encoding for 1 byte tag");
+                heapless::Vec::try_from([b3].as_slice()).unwrap()
+            } else {
+                debug_assert_eq!(
+                    b3 & 0b11111,
+                    0b11111,
+                    "Invalid encoding for first byte of tag"
+                );
+                debug_assert!(
+                    (0x1F..=0x7F).contains(&b3),
+                    "Invalid encoding for first byte of tag"
+                );
+                heapless::Vec::try_from([b2, b3].as_slice()).unwrap()
+            }
+        } else {
+            debug_assert_eq!(
+                b1 & 0b11111,
+                0b11111,
+                "Invalid encoding for first byte of tag"
+            );
+            debug_assert!(b2 > 0x80);
+            debug_assert!((0x00..0x7F).contains(&b3));
+            heapless::Vec::try_from([b1, b2, b3].as_slice()).unwrap()
+        }
+    }
+}
+
+pub fn get_do<'input>(tag_path: &[Tag], data: &'input [u8]) -> Option<&'input [u8]> {
```

Comment:
What is `do`? Data object? If so, can we have `get_data_boject` and `take_data_object` instead?